File: Dynamic Programming/Hard/LET_no1259_Handshakes That Don't Cross.py
class Solution:
    def numberOfWays(self, num_people: int) -> int:

        mod = 10 ** 9 + 7
        pairs = num_people // 2
        # dp[n] 表示有n對握手
        # dp[n + 1] = dp[0] * dp[n] + dp[1] * dp[n - 1] + ... + dp[n] * dp[0]
        dp = [1] + [0] * pairs
        for i in range(1, pairs + 1):
            for j in range(i):
                dp[i] += dp[j] * dp[i - j - 1] % mod

        return dp[pairs] % mod

File: Dynamic Programming/Hard/LET_no1259_Handshakes That Don't Cross.cpp
class Solution {
public:
    int numberOfWays(int numPeople) {
        const static int mod = 1'000'000'007;
        int shakes = numPeople / 2;
        vector<int> dp(shakes + 1, 0);
        dp[0] = 1;

        for (int pair = 1; pair <= shakes; pair++) {
            for (int sub_pair = 0; sub_pair < pair; sub_pair++) {
                dp[pair] += ((long long) dp[sub_pair] * dp[pair - sub_pair - 1]) % mod;
                dp[pair] %= mod;
            }
        }

        return dp[shakes];
    }
};